Add horizon finder into BBH executable

markscheel opened the pull request:
## Proposed changes

Adds horizon finder into BBH executable.

I don't think CI actually runs this executable, so it should be tested before merging. (thats why I made it a draft)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Add Ah finders to BinaryBlackHole executable.

wthrowe commented:
I've managed to write a control system measurement based on this that at least compiles.  (I don't have the data to run it either.)  I think it should be enough for at least the systems that only use the centers.

markscheel commented:
Thanks @wthrowe ! I don't have the data to run this either, so I was hoping that @geoffrey4444 would check out my branch and try it.

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:
```diff
@@ -176,6 +179,92 @@ struct EvolutionMetavars {
+    using compute_vars_to_interpolate = ah::ComputeHorizonVolumeQuantities;
+    // Observe horizon in grid frame too.
+    using tags_to_observe = tmpl::list<
+        StrahlkorperGr::Tags::AreaCompute<::Frame::Grid>,
```
<optional> Since you typically want to do the same thing on AhA and AhB, would be nice to define some aliases once and then use them for both AhA and AhB, wherever aliases in AhA and AhB structs are exactly the same

geoffrey4444 commented on tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:
```diff
@@ -194,6 +204,30 @@ Observers:
+      Center: [*XCoordA, 0.0, 0.0]
+    FastFlow: &DefaultFastFlow
+      Flow: Fast
+      Alpha: 1.0
```
Shouldn't you set these options for AhB as well?

markscheel replied:
Yes, they are set.  Thats what the `&DefaultFastFlow` and the later `*DefaultFastFlow` does.

geoffrey4444 commented on tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:
```diff
@@ -194,6 +204,30 @@ Observers:
+      DivergenceIter: 5
+      MaxIts: 100
+    Verbosity: Verbose
+  AhB:
```
Why are Alpha, Beta, etc. not options to set for AhB?

markscheel replied:
They are.  That's what `*DefaultFastFlow` does.  Look up "Yaml aliases" and "Yaml anchors".

geoffrey4444 reviewed (changes requested):
I have some questions about the code. Also, when I tried running it on a BBH (`ocean:/home/geoffrey/aps2021/inspiral_size_attempt/bbh_exec_attempt/ah/fail1`), I got an error from a root finder complaining about bracketing:

~~~
Line: 61 of /home/geoffrey/Codes/spectre/spectre/src/NumericalAlgorithms/RootFinding/RootBracketing.hpp
Function: std::array<double, 4> RootFinder::bracketing_detail::bracket_by_contracting(const std::vector<double> &, const std::vector<std::optional<double> > &, const Functor &, const size_t) [Functor = (lambda at /home/geoffrey/Codes/spectre/spectre/src/NumericalAlgorithms/RootFinding/RootBracketing.hpp:334:9)]
Too many iterations in bracket_by_contracting. Either refine the initial range/guess or increase max_level.
############ ERROR ############
~~~

Maybe it's trying to invert a map at a place where it's not invertible?

geoffrey4444 commented:
Yes, CI only tests that the input file options work, not that the executable runs successfully

markscheel commented:
Fixup commit adds aliases for things common to AhA and AhB.

Executable should now work provided #3594 is also used; the bracketing error that @geoffrey4444 mentions above is fixed by #3594 .

geoffrey4444 commented:
I will test this again and let you know if it works for me

geoffrey4444 reviewed (changes requested):
I rebased this on latest develop, and I verified that it the AH finder now works successfully!

I'm happy to approve, but first please rebase on latest develop, and verify that the test input file works? In the current version, the domain is out of date. I put the input file I used successfully here:

~~~
wheeler:/home/geoffrey/temp/BinaryBlackHole.yaml
ocean:/home/geoffrey/aps2021/inspiral_bbh_exec_attempt/spectre_run/v2ah/Run/BinaryBlackHole.yaml
~~~

Edit: please squash

markscheel committed:
Update BinaryBlackHole.yaml.

This is made to match Geoffrey's input file, except I removed
commented-out stuff and did not add explicit pathnames.

markscheel commented:
Squashed and rebased, and added another commit that updates the input file.

@geoffrey4444 In your example input file there were lines in EventsAndTriggers that were commented out; when I changed the input file I just removed those.

markscheel commented:
The failing test is Unit.PointwiseFunctions.AnalyticSolutions.Elasticity.HalfSpaceMirror which was not touched in this PR.

nilsvu commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:
```diff
@@ -177,6 +180,73 @@ struct EvolutionMetavars {
+                 StrahlkorperGr::Tags::SpinFunctionCompute<::Frame::Grid>>,
+      horizons_tags_to_observe>;
+
+  struct AhA {
```
(optional) AFAICT the `AhA` and `AhB` classes are identical. How about this:

```cpp
template <typename Label>
struct ApparentHorizon { ... };
struct AhA {};
struct AhB {};
using interpolation_target_tags = tmpl::list<ApparentHorizon<AhA>, ApparentHorizon<AhB>>;
```

markscheel replied:
If I do this, then the code that determines the input file options would need to change too.  Currently the `AhA` in the input file is the same as the InterpolationTargetTag, so if the InterpolationTargetTag changes to `ApparentHorizon<AhA>` then you would need to say something like `ApparentHorizon<AhA>` in the input file (if that would work), or change the code for getting the input file tag from the InterpolationTargetTag.

It seems simpler to me if the name in the input file is the same as the name of the tag in the metavariables.

nilsvu replied:
I think the code that determines the input file name should use `Options::name<>()`, then it's just a matter of defining a static `name()` function. I can't find the place in the code where the input file name is determined though to check what it does. Either way, I'm fine leaving this as-is.

wthrowe replied:
I did basically that in #3620: https://github.com/sxs-collaboration/spectre/pull/3620/files#diff-48693954adbf700f151d18b3160a43b05928950c7e46e12579da1142708f6ea4R54 (`Interpolator` is an interpolation target tag.)

nilsvu replied:
@wthrowe are you requesting a change here?

nilsvu commented on tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:
```diff
@@ -190,6 +190,16 @@ EventsAndTriggers:
         InterpolateToMesh: None
         CoordinatesFloatingPointType: Double
         FloatingPointTypes: [Double]
+  ? Slabs:
+      EvenlySpaced:
+        Interval: 1
+        Offset: 0
+  : - AhA
+  ? Slabs:
+      EvenlySpaced:
+        Interval: 1
+        Offset: 0
+  : - AhB
```
```suggestion
  ? Slabs:
       EvenlySpaced:
         Interval: 1
         Offset: 0
   : - AhA
     - AhB
```

nilsvu commented on tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:
```diff
@@ -136,27 +136,9 @@ Filtering:
     Alpha: 36.0
     HalfPower: 210
-    DisableForDebugging: false
+    DisableForDebugging: true
```
This says "for debugging". Did you intend to change this?

markscheel replied:
I could change it;  @geoffrey4444 wanted me to use his input file, and this was present in his input file.

geoffrey4444 replied:
I no longer recommend filtering for BBH runs, since it doesn't seem necessary in my tests, but I'm unsure whether to remove it altogether, so I prefer disabling in yaml for now

nilsvu commented on tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:
```diff
@@ -136,27 +136,9 @@ Filtering:
-          - Cfl:
-              SafetyFactor: 0.8
-          - Increase:
-              Factor: 1.2
```
Just for my information, why did you remove the change in slab size?

markscheel replied:
I removed it because @geoffrey4444 did;  I just copied his input files that he said that worked.

geoffrey4444 replied:
You can revert and squash this change...I was doing fixed time step runs, but I think this adaptive slab chooser is fine for the test input file

nilsvu reviewed (changes requested):
Please squash directly if you end up making changes.

nilsvu commented:
you can just squash this @markscheel :D

markscheel commented:
Made an input file fix and squashed.  Some other input file things are a result of me copying a working input file from @geoffrey4444 , so perhaps the slab size and the debugging flag can be restored?

nilsvu commented:
@geoffrey4444 could you comment briefly on the input file comments above?

geoffrey4444 commented:
> @geoffrey4444 could you comment briefly on the input file comments above?

I did above :) please let me know if you need more input from me!

markscheel commented:
I put back the ChangeSlabSize stuff in the input file and squashed.
